samza-kafka/src/test/scala/org/apache/samza/system/kafka/TestKafkaSystemProducer.scala

```diff
@@ -165,52 +162,383 @@ class TestKafkaSystemProducer {
+
+    // Subsequent sends
+    producer.send("test1", msg4) // Should succeed because the producer recovered.
+    producer.send("test2", msg5) // Second source should also not have any error.
+
+    // Flushes
+    producer.flush("test2") // Should not throw for test2
+    producer.send("test2", msg6) // Should still work after flush
+
+    producer.flush("test1") // Should not throw for test1 either
+    producer.send("test1", msg7)
+
+    assertEquals(6, mockProducer.getMsgsSent) // every message except the one with the error should get sent
+    producer.stop()
+  }
+
+  @Test
+  def testKafkaProducerWithNonFatalExceptionsMultipleSourcesDroppingExceptions {
+    val msg1 = new OutgoingMessageEnvelope(systemStream, "a".getBytes)
+    val msg2 = new OutgoingMessageEnvelope(systemStream, "b".getBytes)
+    val msg3 = new OutgoingMessageEnvelope(systemStream, "c".getBytes)
+    val msg4 = new OutgoingMessageEnvelope(systemStream, "d".getBytes)
+    val msg5 = new OutgoingMessageEnvelope(systemStream, "e".getBytes)
+    val producerMetrics = new KafkaSystemProducerMetrics()
+
+    val mockProducer = new MockKafkaProducer(1, "test", 1)
+    val producer = new KafkaSystemProducer(systemName = "test",
+      getProducer = () => {
+        mockProducer.open() // A new producer is never closed
+        mockProducer
+      },
+      metrics = producerMetrics,
+      dropProducerExceptions = true)  // Here's where we enable exception dropping.
+    producer.register("test1")
+    producer.register("test2")
+    producer.start()
+
+    producer.send("test1", msg1)
+    producer.send("test2", msg2)
+    mockProducer.setErrorNext(true, false, new SerializationException())
+    val sendException = intercept[SystemProducerException] {
+      producer.send("test1", msg3) // User-thread exception
+    }
+    assertTrue(sendException.getCause.isInstanceOf[SerializationException])
+    assertFalse(mockProducer.isClosed)
+    assertEquals("Should NOT have created a new producer", 1, mockProducer.getOpenCount)
+
+    producer.send("test1", msg3) // Should be able to resend msg3
+    producer.send("test2", msg4) // Second source should not be affected
+
+    producer.flush("test1") // Flush should be unaffected
+
+    producer.send("test1", msg5) // Should be able to send again after flush
+
+    assertEquals(5, mockProducer.getMsgsSent) // only the messages before the error get sent
     producer.stop()
   }
 
   @Test
   def testKafkaProducerFlushMsgsWhenStop {
-    val msg1 = new OutgoingMessageEnvelope(new SystemStream("test", "test"), "a".getBytes)
-    val msg2 = new OutgoingMessageEnvelope(new SystemStream("test", "test"), "b".getBytes)
-    val msg3 = new OutgoingMessageEnvelope(new SystemStream("test", "test"), "c".getBytes)
+    val msg1 = new OutgoingMessageEnvelope(systemStream, "a".getBytes)
+    val msg2 = new OutgoingMessageEnvelope(systemStream, "b".getBytes)
+    val msg3 = new OutgoingMessageEnvelope(systemStream, "c".getBytes)
```

Comment:
It would be nice to use the startDelayedSendThread mechanism to test race conditions between flush() w/ error against send() as well.

Reply:
The tests above test this case but with more determinism. It's part of the reason they're so long. :-)